File: examples/ciflypy-examples/ciflypy_examples/parent_aid.py
import ciflypy as cifly
import ciflypy_examples.utils as utils

ruletables = utils.get_ruletable_path()

poss_desc_table = cifly.Ruletable(str(ruletables / "possible_descendants_cpdag.txt"))
not_amenable_table = cifly.Ruletable(str(ruletables / "not_amenable_cpdag.txt"))
forb_path_conn_table = cifly.Ruletable(str(ruletables / "forbidden_path_connected_cpdag.txt"))
non_causal_conn_table = cifly.Ruletable(str(ruletables / "non_causal_connected_cpdag.txt"))

def poss_desc(g, x):
    return set(cifly.reach(g, {"X": x, "W": []}, poss_desc_table))

def not_amenable(g, x):
    return set(cifly.reach(g, {"X": x}, not_amenable_table))

def forbidden(g, x, W):
    return set(cifly.reach(g, {"X": x, "W": W}, forb_path_conn_table))

def non_causal_connected(g, x, W):
    return set(cifly.reach(g, {"X": x, "W": W}, non_causal_conn_table))

def not_amenable_not_adjustment(g, x, W):
    nam = not_amenable(g, x)
    return nam, nam.union(forbidden(g, x, W)).union(non_causal_connected(g, x, W))

def parent_aid(p, g_true, g_guess):
    parents = [[] for _ in range(p)]
    for u, v in g_guess["-->"]:
        parents[v].append(u)
    g_guess_parsed = cifly.Graph(g_guess, poss_desc_table)
    g_true_parsed = cifly.Graph(g_true, poss_desc_table)
    mistakes = 0

    for x in range(p):
        pt = set(parents[x])
        not_amenable_guess = not_amenable(g_guess_parsed, x)
        desc_true = poss_desc(g_true_parsed, x)
        not_amenable_true, not_adjustment_true = not_amenable_not_adjustment(g_true_parsed, x, pt)

        for y in range(p):
            if y == x:
                continue
            if y in pt:
                if y in desc_true:
                    mistakes += 1
            elif y in not_amenable_guess:
                if y not in not_amenable_true:
                    mistakes += 1
            else:
                if y in not_adjustment_true:
                    mistakes += 1

    return mistakes

File: examples/ciflyr-examples/R/parentAid.R
library("ciflyr")
library("here")

source(here("R", "utils.R"))

ruletables <- getRuletablePath()
possDescendants <- parseRuletable(file.path(ruletables, "possible_descendants_cpdag.txt"))
notAmenable <- parseRuletable(file.path(ruletables, "not_amenable_cpdag.txt"))
forbiddenPath <- parseRuletable(file.path(ruletables, "forbidden_path_connected_cpdag.txt"))
nonCausal <- parseRuletable(file.path(ruletables, "non_causal_connected_cpdag.txt"))

vectorToIndicator <- function(p, l) {
	b <- rep(FALSE, p)
	b[l] <- TRUE
	return (b)
}

reachPossDescendants <- function(p, g, t) {
	return (vectorToIndicator(p, reach(g, list("X" = t, "W" = c()), possDescendants)))
}

reachNotAmenable <- function(p, g, t) {
	return (vectorToIndicator(p, reach(g, list("X" = t), notAmenable)))
}

reachForbiddenPath <- function(p, g, t, Z) {
	return (vectorToIndicator(p, reach(g, list("X" = t, "Z" = Z), forbiddenPath)))
}

reachNonCausal <- function(p, g, t, Z) {
	return (vectorToIndicator(p, reach(g, list("X" = t, "Z" = Z), nonCausal)))
}

notAmenablenotAdjustment <- function(p, g, t, Z) {
	nam <- reachNotAmenable(p, g, t)
	forb <- reachForbiddenPath(p, g, t, Z)
	ncau <- reachNonCausal(p, g, t, Z)

	nad <- rep(FALSE, p)
	nad[nam] <- TRUE
	nad[forb] <- TRUE
	nad[ncau] <- TRUE

	return (list("nam" = nam, "nad" = nad))
}

parentAid <- function(p, gTrue, gGuess) {
	gTrueParsed <- parseGraph(gTrue, possDescendants)
	gGuessParsed <- parseGraph(gGuess, possDescendants)

	paList <- parents(p, gGuess)

	mistakes <- 0
	for (t in seq(p)) {
		pa <- vectorToIndicator(p, paList[[t]])
		namGuess <- reachNotAmenable(p, gGuessParsed, t)
		desTrue <- reachPossDescendants(p, gTrueParsed, t)
		namnadTrue <- notAmenablenotAdjustment(p, gTrueParsed, t, paList[[t]])

		for (y in seq(p)) {
			if (y == t) {
				next
			}
			if (pa[y]) {
				if (desTrue[y]) {
					mistakes <- mistakes + 1
				}
			} else if (namGuess[y]) {
				if (!namnadTrue$nam[y]) {
					mistakes <- mistakes + 1
				}
			} else {
				if (namnadTrue$nad[y]) {
					mistakes <- mistakes + 1
				}
			}
		}
	}
	return (mistakes)
}
